datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaBasedConnectorTask.java

```diff
@@ -257,18 +266,26 @@ protected void rewindAndPausePartitionOnException(TopicPartition srcTopicPartiti
       seekToLastCheckpoint(Collections.singleton(srcTopicPartition));
       partitionRewound = true;
     } catch (Exception e) {
-      _logger.error("Partition rewind failed due to ", e);
+      _logger.error(String.format("Partition rewind for %s failed due to ", srcTopicPartition), e);
     }
     if (_pausePartitionOnError && (!partitionRewound || !containsTransientException(ex))) {
-      // if doesn't contain DatastreamTransientException and it's configured to pause partition on error conditions,
-      // add to auto-paused set
+      // If partition rewind failed or the exception is not of type DatastreamTransientException and it is configured to
+      // pause partition on error conditions, add it to the auto-paused set
       _logger.warn("Adding source topic partition {} to auto-pause set", srcTopicPartition);
       _autoPausedSourcePartitions.put(srcTopicPartition,
           PausedSourcePartitionMetadata.sendError(start, _pauseErrorPartitionDuration, ex));
       _taskUpdates.add(DatastreamConstants.UpdateType.PAUSE_RESUME_PARTITIONS);
     }
   }
 
+  protected void rewindAndPausePartitionsOnSendException() {
+    // For all topic partitions which have seen send exceptions, attempt to rewind them to the last checkpoint
+    synchronized (_sendFailureTopicPartitionExceptionMap) {
+      _sendFailureTopicPartitionExceptionMap.forEach(this::rewindAndPausePartitionOnException);
```

Comment:
You should check if the topic-partition is already in the pause list and skip the rewind process.  Reason: suppose for a topic-partition, 10 records are polled and got 1 callback with error before next poll, this topic partition will be rewind and paused and poll will continue. Suppose the other 9 callbacks are coming with delay and come before next 2-3 poll iterations, we will do this rewind multiple-times, even though the partition is already paused and rewind. We should try to skip the same no-op  operation multiple times.

Reply:
Yeah so I did think that, but decided against it. When we add a TopicPartition to the auto-pause list on send failure, we can do it for two reasons: a) the send exception is not transient, b) the rewind failed. For (b) I think it is safer to potentially attempt the rewind on the next send failure if possible, in case the first rewind failed for some transient reason. When we resume TopicPartitions from the auto-pause list, we don't attempt to rewind them then.

The maximum number of rewinds we'll do for a TopicPartition = number of polls between flushes, since on each poll/flush we'll rewind the TopicPartition only once in spite of receiving multiple send callbacks.

We can potentially differentiate between rewind failure vs. non-transient exception and use that to decide whether to rewind in the future. This may need more thought though. The current code is no worse than the existing code in terms of doing no-op rewinds. What do you think? Okay to leave this as is for now?

Reply:
Do you want to mark it as TODO then with the explanation?

Reply:
Sure, will do, and I'll open a ticket. Will not leave this change hanging forever.

Reply:
Can you add TODO and the explanation so that the thought is not lost in the process.

Reply:
done